src/main/java/net/minecraftforge/client/gui/config/widgets/ConfigGuiWidget.java

```diff
@@ -0,0 +1,740 @@
+     */
+    public static class LongWidget extends NumberWidget<Long>
+    {
+        public static ConfigGuiWidgetFactory FACTORY = (value, valueSpec, valueManager, spec, name) ->
+        {
+            if (value instanceof ForgeConfigSpec.LongValue)
+            {
+                return new LongWidget(valueSpec, valueManager, spec, name);
+            }
+
+            throw new IllegalArgumentException("The given config value for path: " + String.join(".",
+                    value.getPath()) + " is not a long value!");
+        };
+
+        public LongWidget(final ForgeConfigSpec.ValueSpec valueSpec,
+                          final ValueManager valueManager, SpecificationData specificationData, final Component name)
+        {
+            super(valueSpec, valueManager, specificationData, name, Long::parseLong);
+        }
+    }
+
+    /**
+     * Default implementation for a {@link ConfigGuiWidget} for a {@link ForgeConfigSpec.DoubleValue}.
+     */
+    public static class DoubleWidget extends NumberWidget<Double>
+    {
+        public static ConfigGuiWidgetFactory FACTORY = (value, valueSpec, valueManager, spec, name) ->
+        {
+            if (value instanceof ForgeConfigSpec.DoubleValue)
+            {
+                return new DoubleWidget(valueSpec, valueManager, spec, name);
+            }
+
+            throw new IllegalArgumentException("The given config value for path: " + String.join(".",
+                    value.getPath()) + " is not a Double value!");
+        };
+
+        public DoubleWidget(final ForgeConfigSpec.ValueSpec valueSpec, final ValueManager valueManager,
+                            SpecificationData specificationData, final Component name)
+        {
+            super(valueSpec, valueManager, specificationData, name, Double::parseDouble);
+        }
+    }
+
+    /**
+     * Default implementation for a {@link ConfigGuiWidget} for a {@link ForgeConfigSpec.ConfigValue} which references a {@link String}.
+     */
+    public static class TextWidget extends ConfigGuiWidget
+    {
+        @SuppressWarnings("unchecked")
+        public static ConfigGuiWidgetFactory FACTORY = (value, valueSpec, valueManager, spec, name) ->
+        {
+            try
+            {
+                // Try to cast it to something that we can handle to verify it is of the right type.
+                // We should really just provide a spec for it but meh for now this suffices.
+                final ForgeConfigSpec.ConfigValue<? extends String> stringValue = (ForgeConfigSpec.ConfigValue<? extends String>) value;
+                return new TextWidget(valueSpec, valueManager, spec, name);
+            }
+            catch (ClassCastException ex)
+            {
+                throw new IllegalArgumentException("The given config value for path: " + String.join(".",
+                        value.getPath()) + " is not a Double value!", ex);
```

Comment:
Copy paste error? As we aren't expecting a double value here (unless I am massively mistaken)